### DISCUSSION TITLE
Add getter methods to StackItem that allow omitting the actual stack item type

### DISCUSSION THREAD

**Discussion by csmuller** [neow3j/neow3j#331] [Ideas]
Currently it is rather cumbersome to fetch a stack item from invocation result object, e.g., `response.getInvocationResult().getStack().get(0).asByteString().getAsHexString()`.
This also implies that the developer knows exactly what `StackItemType` to expect from the response. 

The idea is to add abstract methods to the `StackItem` class that are implemented differently by the subclasses and abstract away the actual type of the stack item. The developer would be able to do the following.
```
response.getInvocationResult().getStack().get(0).asHexString()
```
She doesn't need to know the exact type of the stack item. It could be `Buffer`, `ByteString`, `Integer` or `Boolean`, they are all representable as a hex string and would implement the `asHexString()` method. Other types like `Array` or `Map` would throw an exception because no reasonable hex string can be returned for them.

This approach is also a solution of the problem mentioned in the issue #272. By having a method `StackItem.asBoolean()`, neow3j would return a boolean even if the actual stack item type is `Integer`.

What do you think @mialbu & @gsmachado?

**Comment by gsmachado** [comment]
@csmuller this is reasonable, yes.

But I would prefer to include additional convenience methods, like:

```
response.getInvocationResult().getStack().getFirst().asHexString()
```

```
response.getInvocationResult().getFirstItemFromStack().asHexString()
```

What do you think?

**Reply by gsmachado** [reply]
Hmmmm.... how a `Boolean` would be represented in the `.asHexString()` then? Like `01` or `00`?

Could we also include other methods, such as `.asBooleanString()` or something like that?

Let me know your thoughts.

**Reply by csmuller** [reply]
I also see some value in convenience methods like you mentioned. Getting the first stack item with `.getStack().get(0)` is not nice, so something of the sort as you mentioned is an improvement.

> Hmmmm.... how a `Boolean` would be represented in the `.asHexString()` then? Like `01` or `00`?

Yes, `01` and `00` make sense in my opinion. But, now i think the usage of these convenience methods is a bit more restricted than how I mentioned them in the beginning of this discussion.  
`asHexString()` is mostly useful for `Buffer` and `ByteString` stack items.   
`asBollean()` is mostly useful for `Boolean` and `Integer` stack items.  
The addition of these methods on `StackItem` would solve the problem of return types not matching the types denoted in the contract manifest. But getting a `Boolean` stack item as a hex string doesn't make much sense. In that case I would throw an exception to notify the developer that the returned type is too different from what she expected.

**Reply by gsmachado** [reply]
> But, now i think the usage of these convenience methods is a bit more restricted than how I mentioned them in the beginning of this discussion.

True.

> In that case I would throw an exception to notify the developer that the returned type is too different from what she expected.

I would do the same. It's better to provide these convenience methods even if they can be "misused". If the developers wrongly use it, we can simply make them aware of what/why it's wrong.

My final saying, is: we really need to hide `.getStack().get(0)` as much as we can. We don't have to remove from the API interface... but we should present neow3j SDK without using this.

**Reply by gsmachado** [reply]
Do we have an issue to address this already?

**Reply by csmuller** [reply]
Yes, I opened #337.
